File: SEGUNDO/PDOO/PRACTICAS/PRACTICA2/java/deepSpace/src/deepspace/Damage.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package deepspace;

import java.util.ArrayList;
import java.util.Iterator;

/**
 *
 * @author joarpe
 */
public abstract class Damage {
    private int nShields; //nWeapons;
    //private ArrayList<WeaponType> weapons;
    
    Damage (int s) {
        nShields = s;
        //nWeapons = w;
        //weapons = new ArrayList<>();
    }
    
    /*Damage (ArrayList<WeaponType> wl, int s) {
        nShields = s;
        //nWeapons = -1;
        //weapons = new ArrayList (wl);
    }*/
    
    /*Damage (Damage d) {
        nShields = d.getNShields();
        nWeapons = d.getNWeapons();
        weapons = new ArrayList(d.getWeapons());
    }*/
    
    public abstract Damage copy ();
    
    public abstract DamageToUI getUIversion ();
    
    /*private int arrayContainsType (ArrayList<Weapon> w, WeaponType t) {
        return w.indexOf(t);
    }*/
    
    public int adjustShields(ArrayList<ShieldBooster> s) {
        int nS = nShields;
        if (s.size()<nShields) 
            nS = s.size();
        
        return nS;
    }
    
    public abstract Damage adjust (ArrayList<Weapon> w, ArrayList<ShieldBooster> s); //{
        //int nS = nShields;
        //int nW = nWeapons;
        //ArrayList<WeaponType> weapon_type;
        //weapon_type = new ArrayList (weapons);
        
        //if (s.size() < nShields) 
                //nS=s.size();
        
        /*if(nWeapons!=-1) {
            if (w.size() < nWeapons)
                nW=w.size();
            return new Damage(nW, nS);
        }
        else {
            ArrayList<Weapon> weap;
            weap = new ArrayList(w);
            
            for (int i=0; i<weapons.size(); ++i) {
                boolean found = false;
                
                for (int j=0; (j<weap.size()) && !found; ++j) {
                    if (weapons.get(i)==weap.get(j).getType()) {
                        found = true;
                        weap.remove(j);
                    }
                }
                
                if (!found) {
                    weapon_type.remove(weapons.get(i));
                }
            }
            
            return new Damage(weapon_type, nS);
        }*/ 
        
        //return nS;
                
    //}
    
    public abstract void discardWeapon (Weapon w);
    
    public void discardShieldBooster () {
        if (nShields > 0) {
            nShields--;
        }
    }
    
    public boolean hasNoEffect () {
        return (nShields == 0); //&& (nWeapons == 0 || weapons.size()==0));
    }
    
    public int getNShields () {
        return nShields;
    }
    
    /*public int getNWeapons () {
        return nWeapons;
    }
    
    public ArrayList<WeaponType> getWeapons () {
        return weapons;
    }*/
    
    public String toString () {
        return ""+nShields;
    }
}

File: SEGUNDO/PDOO/PRACTICAS/PRACTICA2/ruby/Damage.rb
#encoding:utf-8

require_relative 'DamageToUI.rb'
require_relative 'WeaponType.rb'

module Deepspace
    class Damage
        private_class_method :new

        def initialize(s)
            @nShields = s
        end
=begin
        def self.newNumericWeapons(w, s)
            aux = Array.new
            new(w,s,aux)
        end

        def self.newSpecificWeapons(wl, s)
            new(@@NOT_USED, s, wl)
        end

        def self.newCopy(d)
            if d.nWeapons == -1
                return newSpecificWeapons(d.weapons, d.nShields)
            else
                return newNumericWeapons(d.nWeapons, d.nShields)
            end
        end

        private
        def arrayContainsType (w, t)
            for i in(0..w.size-1) do
                if (w.at(i).type == t)
                    return i
                end
            end
            return -1
        end
=end
        def adjustShields (s)
            nS=@nShields

            if (s.size<@nShields)
                nS=s.size
            end

            return nS
        end

=begin
        def adjust (w, s)
            nW=@nWeapons
            nS=@nShields
            weapon_type=Array.new(@weapons)
            #weapons_type=WeaponType.newCopy(@weapons)

            if (s.size()<nS)
                nS=s.size()
            end
            
            if(@nWeapons!=-1)
                if (w.size < @nWeapons)
                    nW=w.size
                end
                self.class.newNumericWeapons(nW, nS)
            else
                weap=Array.new(w)
                i=0
                while i<@weapons.size
                    found = false

                    j=0
                    while j<weap.size && !found
                        if @weapons.at(i)==weap.at(j).type
                            found = true
                            weap.delete_at(j)
                        end
                        j=j+1
                    end

                    if !found
                        weapon_type.delete(@weapons.at(i))
                    end
                    i=i+1
                end
                self.class.newSpecificWeapons(weapon_type, nS)
            end
        end

        def discardWeapon (w)
            if (@nWeapons == -1)
                if (@weapons.length != 0)
                    @weapons.delete(w.type)
                end
            elsif @nWeapons >0
                @nWeapons=@nWeapons-1
            end
        end
=end

        def discardShieldBooster
            if (@nShields>0)
                @nShields = @nShields-1
            end
        end

        def hasNoEffect
            (@nShields==0)
        end

        def nShields
            @nShields
        end

=begin
        def nWeapons
            @nWeapons
        end

        def weapons
            @weapons
        end
=end

        def to_s
            ""+@nShields
        end

    end
end